### DISCUSSION TITLE
Securing deployment supply chain with Flux V2 - validating build-time signatures and signed ACR images?

### DISCUSSION THREAD

**Discussion by alukyan** [fluxcd/flux2#3956] [General]
Hello,

I am an architect on large scale infrastructure project (>1000 K8s clusters and growing) and I am looking if we can leverage Flux V2 as a deployment agent in our clusters instead of our proprietary solution. It is great to see continuous attention to security issues and in the roadmap. However, I have a few questions to see how Flux V2 can be used with our security requirements.

1. One of our requirements for the secure supply chain is to make sure that deployed artifacts have not been tampered with since the time they were produced by the secure build pipelines. This is because application release pipelines in distributed multitenant environment may contain a few attack vectors that allow to inject malicious code or configuration in them. In our current deployment system this requirement is satisfied by the build system signing (GPG signature) of a build manifest - a file with the list of all artifact files and their SHA256 hashes. When the artifacts are pulled into the target cluster we need to (a) validate that the build manifest is still valid by checking its signature with trusted GPG public key, and (b) artifact files SHA256 hashes match the ones in the validated build manifest.

2. If we use Azure Container Registry as an [OCI registry](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-oci-artifacts) we need to validate that container images and OCI artifacts are signed.

I could not find any support for (1), but if I were to extend the Flux to support this requirement the most logical choice to do this is in the Source Controller. Another, less favorable option would be to create a custom Source Controller, but this is a more expensive option. Do you have any recommendations on how to proceed here?

For (2) I see that Azure Container Registry supports signing container images and OCI artifacts [using Notary](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-tutorial-sign-build-push) and Flux V2 supports verification of [Cosign signatures](https://fluxcd.io/flux/components/source/ocirepositories/#verification). I could not find any information if those signing formats are compatible. Do you know the answer here? If not, what is your strategy for supporting container image signing formats offered by large cloud providers?

**Comment by stefanprodan** [comment]
1) We support Git commit signature verification using OpenPGP, docs here https://fluxcd.io/flux/components/source/gitrepositories/#verification

2) Notation signature’s verification is being implemented here https://github.com/fluxcd/source-controller/pull/1075

**Comment by alukyan** [comment]
Unfortunately, Git cannot be used as configuration source in compliant production environments. Those environments require separation of production and development for all identities - user and system.

And Git commit signature verification is not the same as build time signatures. Git commit applies to the source code and build signatures apply to the artifacts produced from the source code via the build process. Build and release systems are common attack vectors on integrity of the supply chain. If your company has penetration testing team, they commonly use this vector of attack. 

Great to see the support for Notation being added!
